t/apicast-policy-limits.t

```diff
@@ -0,0 +1,301 @@
+"POST /test \n" . $ENV{LARGE_BODY}
+--- response_body
+Payload Too Large
+--- error_code: 413
+--- no_error_log
+[error]
+
+
+=== TEST 3: Request body size smaller than the limit
+-- ONLY
+--- backend
+  location /transactions/authrep.xml {
+    content_by_lua_block {
+      local expected = "service_token=token-value&service_id=42&usage%5Bhits%5D=2&user_key=value"
+      require('luassert').same(ngx.decode_args(expected), ngx.req.get_uri_args(0))
+    }
+  }
+--- configuration
+{
+  "services": [
+    {
+      "id": 42,
+      "backend_version":  1,
+      "backend_authentication_type": "service_token",
+      "backend_authentication_value": "token-value",
+      "proxy": {
+        "proxy_rules": [
+          { "pattern": "/", "http_method": "POST", "metric_system_name": "hits", "delta": 2 }
+        ],
+        "policy_chain": [
+          {
+            "name": "apicast.policy.limits",
+            "configuration": {
+              "request": 10000000,
+              "response": 0
+            }
+          },
+          { "name": "apicast.policy.upstream",
+            "configuration":
+              {
+                "rules": [ { "regex": "/", "url": "http://test:$TEST_NGINX_SERVER_PORT" } ]
+              }
+          }
+        ]
+      }
+    }
+  ]
+}
+--- upstream
+  location / {
+     content_by_lua_block {
+       ngx.say('yay, api backend');
+     }
+  }
+
+--- request eval
+"POST /test \n" . $ENV{LARGE_BODY}
+--- response_body
+yay, api backend
+--- error_code: 200
+--- no_error_log
+[error]
+
+
+=== TEST 4: Response limit set to 100 rejects the request
+--- backend
+  location /transactions/authrep.xml {
+    content_by_lua_block {
+      local expected = "service_token=token-value&service_id=42&usage%5Bhits%5D=2&user_key=value"
+      require('luassert').same(ngx.decode_args(expected), ngx.req.get_uri_args(0))
+    }
+  }
+--- configuration
+{
+  "services": [
+    {
+      "id": 42,
+      "backend_version":  1,
+      "backend_authentication_type": "service_token",
+      "backend_authentication_value": "token-value",
+      "proxy": {
+        "proxy_rules": [
+          { "pattern": "/", "http_method": "POST", "metric_system_name": "hits", "delta": 2 }
+        ],
+        "policy_chain": [
+          {
+            "name": "apicast.policy.limits",
+            "configuration": {
+              "request": 0,
+              "response": 100
+            }
+          },
+          { "name": "apicast.policy.upstream",
+            "configuration":
+              {
+                "rules": [ { "regex": "/", "url": "http://test:$TEST_NGINX_SERVER_PORT" } ]
+              }
+          }
+        ]
+      }
+    }
+  ]
+}
+--- upstream
+  location / {
+     content_by_lua_block {
+      local response = ""
+      for var=0,20480 do
+          response = response .. "XXXXXXXXXX"
+      end
+      ngx.header['Content-Length'] = #response
+      ngx.say(response)
+     }
+  }
+
+--- request
+POST /test
+--- error_code: 413
```

Comment:
I think this status code is just for the request size, no the response one.

Reply:
can't follow this one.

Reply:
What I mean is that I think that the response code 413 is used only when the request payload is higher than the limit. However, I'm not sure what the code should be when the _response_ payload returned by the API upstream is higher than the limit.